src/hotspot/share/jfr/periodic/sampling/jfrCPUTimeThreadSampler.cpp

```diff
@@ -0,0 +1,889 @@
+  u4 _max_frames;
+  // error code for the trace, 0 if no error
+  u4 _error;
+  int64_t _sampling_period;
+
+  JfrSampleType _type;
+  JfrTicks _start_time;
+  JfrTicks _end_time;
+  JavaThread* _sampled_thread;
+
+public:
+  JfrCPUTimeTrace(u4 index, JfrAsyncStackFrame* frames, u4 max_frames):
+    _index(index), _frames(frames), _stacktrace(_frames, max_frames),
+    _max_frames(max_frames) {
+
+    }
+
+  JfrAsyncStackFrame* frames() { return _frames; }
+  u4 max_frames() const { return _max_frames; }
+
+  bool successful() const { return _error == NO_ERROR; }
+
+  JfrSampleType type() const { return _type; }
+
+  JfrTicks start_time() const { return _start_time; }
+  void set_end_time(JfrTicks end_time) { _end_time = end_time; }
+  JfrTicks end_time() const { return _end_time; }
+  void set_sampled_thread(JavaThread* thread) { _sampled_thread = thread; }
+  JavaThread* sampled_thread() const { return _sampled_thread; }
+
+  JfrAsyncStackTrace& stacktrace() { return _stacktrace; }
+
+  int64_t sampling_period() const { return _sampling_period; }
+
+  enum SampleError {
+    NO_ERROR = 0,
+    ERROR_NO_TRACE = 1,
+    ERROR_NO_TOPFRAME = 2,
+    ERROR_JAVA_WALK_FAILED = 3,
+    ERROR_NATIVE_WALK_FAILED = 4,
+    ERROR_NO_TOP_METHOD = 5,
+    ERROR_NO_LAST_JAVA_FRAME = 6
+  };
+
+  // Record a trace of the current thread
+  void record_trace(JavaThread* jt, void* ucontext, int64_t sampling_period) {
+    _stacktrace = JfrAsyncStackTrace(_frames, _max_frames);
+    set_sampled_thread(jt);
+    _type = NO_SAMPLE;
+    _error = ERROR_NO_TRACE;
+    _start_time = _end_time = JfrTicks::now();
+    _sampling_period = sampling_period;
+    if (!jt->in_deopt_handler() && !Universe::heap()->is_stw_gc_active())  {
+      ThreadInAsgct tia(jt);
+      if (thread_state_in_java(jt)) {
+        record_java_trace(jt, ucontext);
+      } else if (thread_state_in_native(jt)) {
+        record_native_trace(jt, ucontext);
+      }
+    }
+    _end_time = JfrTicks::now();
+  }
+
+private:
+
+  void record_java_trace(JavaThread* jt, void* ucontext) {
+    _type = JAVA_SAMPLE;
+    JfrGetCallTrace trace(true, jt);
+    frame topframe;
+    if (trace.get_topframe(ucontext, topframe)) {
+      _error = _stacktrace.record_async(jt, topframe) ? NO_ERROR : ERROR_JAVA_WALK_FAILED;
+    }
+  }
+
+  void record_native_trace(JavaThread* jt, void* ucontext) {
+    // When a thread is only attach it will be native without a last java frame
+   _type = NATIVE_SAMPLE;
+    _error = ERROR_NO_TRACE;
+    if (!jt->has_last_Java_frame()) {
+      _error = ERROR_NO_LAST_JAVA_FRAME;
+      return;
+    }
+    frame topframe;
+    if (!jt->pd_get_top_frame_for_signal_handler(&topframe, ucontext, false)) {
+      _error = ERROR_NO_TOPFRAME;
+      return;
+    }
+    frame first_java_frame;
+    Method* method = nullptr;
+    JfrGetCallTrace gct(false, jt);
+    if (!gct.find_top_frame(topframe, &method, first_java_frame)) {
+      _error = ERROR_NO_TOPFRAME;
+      return;
+    }
```

Comment:
Wouldn't we end up with the same top frame if we started at the anchor frame instead?  In fact, the frame we want should always be the immediate caller of the anchor frame.  But I guess we need to check that the anchor frame is walkable.

Reply:
This is not always the case, as far as I know.

Reply:
This is not always the case, as far as I know.

Reply:
It should be ok to start from the anchor frame.

Please see JfrNativeSamplerCallback::call().